Fix buffer incompatibility issue when used inside React (issue/3580)

elmariachi111 opened the pull request:
**NOTE: PR targets #3583**
 
fixes a type issue between UInt8Arrays and Buffer.concat in browsers.

## Description

Please include a summary of the changes and be sure to follow our [Contribution Guidelines](../CONTRIBUTIONS.md).

Optional if an issue is fixed:
Fixes #3580 

## Type of change

- [ x ] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [ x ] I have selected the correct base branch.
- [ x ] I have performed a self-review of my own code.
- [ o ] I have commented my code, particularly in hard-to-understand areas.
- [ o ] I have made corresponding changes to the documentation.
- [ o ] My changes generate no new warnings.
- [ o ] Any dependent changes have been merged and published in downstream modules.
- [ x ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ x ] I ran `npm run test:unit` with success.
- [ x ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ x ] I ran `npm run build-all` and tested the resulting files from `dist` folder in a browser.
- [ x ] I have updated the `CHANGELOG.md` file in the root folder.
- [ - ] I have tested my code on the live network.
- [ x ] I have checked the Deploy Preview and it looks correct.

elmariachi111 committed:
(untested) potential fix for #3580

I haven't even executed this. Maybe it's... already working like that?!
dunno if ES6 spread ops already are allowed here?!

elmariachi111 committed:
added #3580 fix to changelog

ryanio commented:
hey @elmariachi111 thanks so much for looking into this and opening a PR.

do you think you could try to add a test case that fails without the fix?

elmariachi111 commented:
that's what I also thought about because actually: according to all tests that shouldn't happen at all and they all cover that use case. It only occurs in a browser context, on node the decryption just runs fine...  so, if you can point me to the repo location where e2e tests are written, I'd try my very best ;) So something close to https://github.com/ethereum/web3.js/blob/1.x/test/e2e.method.send.js would to best, right? Is the ci picking up everything that starts with `e2e` automatically for e2e testing?

cgewecke commented:
@elmariachi111 Yes! If you put a test in any of those files it will run in Chrome / Firefox.

elmariachi111 committed:
#3580 adds e2e test that covers the incompatible line
changed the Changelog to reflect relation to React

elmariachi111 commented:
ok, added an e2e test that covers these lines. Interesting: The test won't detect that behaviour, since as I in the meanwhile figured out, the error only occurs when running inside React (or inside code that has been bundled) and thereofre cannot be reproduced when ran in plain JS. The core reason is that the "global" `Buffer` object behaves differently in web3 and in React, maybe due to the transitive dependency on `node_modules/node-libs-browser/node_modules/buffer/index.js:399`. I added files to reproduce that behaviour in my demo repo: https://github.com/cod1ng-earth/demo-web3-129-breaks-keystore-decryption

the "fix" I'm proposing here doesn't seem to have adverse effects on anything else, so I think we can go with it without any issues..

elmariachi111 commented:
oh, btw 2, I noticed that the root package-lock.json on the 1.2.9 release hasn't been committed. Since this is my very first contribution I'm quite surprised that I ended up with a changed package-lock.json but I didn't commit it (this is the one on the tag: https://github.com/ethereum/web3.js/blob/v1.2.9/package-lock.json referring to v1.2.6) and that's the one from the 1.x branch (https://github.com/ethereum/web3.js/blob/1.x/package-lock.json) which carries v1.2.9 in its package.json...

cgewecke reviewed (approved):
@elmariachi111 Thanks so much for this! Just great.

Am going to merge this into #3583 for further triage. 

Unfortunately decrypt is not the only affected method and it looks the feross/buffer polyfill might be incompatible with something in ethereumjs-util as well.
